trie-db/src/iterator.rs

```diff
@@ -184,94 +201,85 @@ impl<'a, L: TrieLayout> TrieDBNodeIterator<'a, L> {
-		-> Result<Node<'c>, TrieHash<L>, CError<L>>
-	{
-		let node = L::Codec::decode(&node_data)
-			.map_err(|e| Box::new(TrieError::DecoderError(node_hash.unwrap_or_default(), e)))?;
-		self.trail.push(Crumb {
-			hash: node_hash,
-			status: Status::Entering,
-			node: Rc::new(node.clone().into()),
-		});
-		Ok(node)
-	}
-}
-
-impl<'a, L: TrieLayout> TrieIterator<L> for TrieDBNodeIterator<'a, L> {
-	fn seek(&mut self, key: &[u8]) -> Result<(), TrieHash<L>, CError<L>> {
-		self.trail.clear();
-		self.key_nibbles.clear();
-		let root_node = self.db.root_data()?;
-		let root_hash = self.db.root().clone();
-		self.seek(root_node, Some(root_hash), NibbleSlice::new(key.as_ref()))
-	}
 }
 
 impl<'a, L: TrieLayout> Iterator for TrieDBNodeIterator<'a, L> {
-	type Item = Result<(NibbleVec, Option<TrieHash<L>>, Rc<OwnedNode>), TrieHash<L>, CError<L>>;
+	type Item = Result<(NibbleVec, Option<TrieHash<L>>, Rc<OwnedNode<DBValue>>), TrieHash<L>, CError<L>>;
 
 	fn next(&mut self) -> Option<Self::Item> {
 		enum IterStep<O, E> {
 			YieldNode,
-			Continue,
 			PopTrail,
-			Descend(Result<(DBValue, Option<O>), O, E>),
+			Descend(Result<(OwnedNode<DBValue>, Option<O>), O, E>),
 		}
 		loop {
 			let iter_step = {
-				let b = self.trail.last()?;
-
-				match (b.status.clone(), b.node.as_ref()) {
-					(Status::Entering, _) => IterStep::YieldNode,
-					(Status::Exiting, n) => {
-						match *n {
-							OwnedNode::Empty | OwnedNode::Leaf(_, _) => {},
-							OwnedNode::Extension(ref n, _) =>
-								self.key_nibbles.drop_lasts(n.len()),
-							OwnedNode::Branch(_) => { self.key_nibbles.pop(); },
-							OwnedNode::NibbledBranch(ref n, _) =>
-								self.key_nibbles.drop_lasts(n.len() + 1),
-						}
-						IterStep::PopTrail
-					},
-					(Status::At, &OwnedNode::Extension(ref partial, ref d)) => {
-						self.key_nibbles.append(partial);
-						IterStep::Descend::<TrieHash<L>, CError<L>>(
-							self.db.get_raw_or_lookup(&*d, self.key_nibbles.as_prefix())
-						)
-					},
-					(Status::At, &OwnedNode::Branch(_)) => {
-						self.key_nibbles.push(0);
-						IterStep::Continue
-					},
-					(Status::At, &OwnedNode::NibbledBranch(ref partial, _)) => {
-						self.key_nibbles.append(partial);
-						self.key_nibbles.push(0);
-						IterStep::Continue
-					},
-					(Status::AtChild(i), &OwnedNode::Branch(ref branch))
-					| (Status::AtChild(i), &OwnedNode::NibbledBranch(_, ref branch)) => {
-						if let Some(child) = branch.index(i) {
-							self.key_nibbles.pop();
-							self.key_nibbles.push(i as u8);
-							IterStep::Descend::<TrieHash<L>, CError<L>>(
-								self.db.get_raw_or_lookup(child, self.key_nibbles.as_prefix())
-							)
-						} else {
-							IterStep::Continue
+				let b = self.trail.last_mut()?;
+
+				if let Status::Entering = b.status {
+					// Handle the Status::Entering case separately to avoid constructing the node
+					// unnecessarily.
+					IterStep::YieldNode
+				} else {
+					let node = b.node.node();
```

Comment:
naïve question, can't we just match on NodePlan instead, and only access the fields if they are needed.

Reply:
Nice, didn't think of that! With this, it's a 43% speedup over master!